Fixes SOLR-13539

thomaswoeckinger opened the pull request:
Introduce EmbeddedSolrTestBase for easy and simple integration tests, fixes are included for SOLR-13331 and SOLR-13347.

@gerlowskija Please review

thomaswoeckinger commented:
Found two other bugs when using different codec than JavaBinCodec, will do another PR
@gerlowskija Did you have time to review the PR?

thomaswoeckinger commented:
@ErickErickson How to get reviewer for this PR Jason seems to be busy

ErickErickson commented:
Unfortunately this is an ongoing issue. Usually, people will gently nudge with additional comments on the JIRA.

I’m slammed for time too unfortunately.

> On May 10, 2019, at 1:49 AM, Thomas Wöckinger <notifications@github.com> wrote:
> 
> @ErickErickson How to get reviewer for this PR Jason seems to be busy
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

dsmiley commented:
It'd help to review your changes if you made fewer arbitrary changes, like adding 'final' and changing indentation of javadocs that were fine as they were.
Also, it'd help to summarize why 3 different issues are being fixed in one PR.  Might be just fine but please add info/context to make reviewer's job either or you may not get a review at all.

ErickErickson commented:
There are two things that might help with the arbitrary change issue:

1> On the reviewing side, In InteliJ, I can choose an option “ignore all whitespace”, I’m sure other tools have similar. Doesn’t help with final and the like of course.

2> Again in IntelliJ and (presumably) other tools I can set the autoformat to only reformat changed lines rather than entire files.

FWIW,
Erick

> On May 15, 2019, at 6:49 AM, David Smiley <notifications@github.com> wrote:
> 
> It'd help to review your changes if you made fewer arbitrary changes, like adding 'final' and changing indentation of javadocs that were fine as they were.
> Also, it'd help to summarize why 3 different issues are being fixed in one PR. Might be just fine but please add info/context to make reviewer's job either or you may not get a review at all.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

thomaswoeckinger commented:
The files where formatted with settings generated by 'ant eclipse', so they
where wrong before, i can remove the final modifier which i personally
prefer.

Some additional inflammations about the issues:

The issues are related to each other due to the change from JavaBinCodec to
use UTF8CharSequence instead of String, which leads to a collection.remove
calls which are useless  because their type does not match.

The issue was introduced in version 7.7.x and initially detected on our
side by integration tests. I want to make sure that this issues will not
occur again in the future so i added missing unit tests and refactored the
test base to allow easy exchange of the used codec.
The test written for SOLR-13331 shows up two additional issues described by
SOLR-13347 and SOLR-11841. Also some unexpected internal state where caused
by the EmbeddedSolrServer which i corrected to (return null stream instead
of empty).

So i will remove the final modifiers and push again!

Thx for your time!

Best,
Tom

Erick Erickson <notifications@github.com> schrieb am Mi., 15. Mai 2019,
19:56:

> There are two things that might help with the arbitrary change issue:
>
> 1> On the reviewing side, In InteliJ, I can choose an option “ignore all
> whitespace”, I’m sure other tools have similar. Doesn’t help with final and
> the like of course.
>
> 2> Again in IntelliJ and (presumably) other tools I can set the autoformat
> to only reformat changed lines rather than entire files.
>
> FWIW,
> Erick
>
> > On May 15, 2019, at 6:49 AM, David Smiley <notifications@github.com>
> wrote:
> >
> > It'd help to review your changes if you made fewer arbitrary changes,
> like adding 'final' and changing indentation of javadocs that were fine as
> they were.
> > Also, it'd help to summarize why 3 different issues are being fixed in
> one PR. Might be just fine but please add info/context to make reviewer's
> job either or you may not get a review at all.
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub, or mute the thread.
> >
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/lucene-solr/pull/665?email_source=notifications&email_token=ACKCGQE2FNGKHHJ6EP6WVA3PVRFFBA5CNFSM4HLKBL4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVPOJZQ#issuecomment-492758246>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACKCGQCSSE62WLOMMUEHKLLPVRFFBANCNFSM4HLKBL4A>
> .
>

janhoy commented:
Yes, please do only feature related changes in the PR, then you can file another JIRA for other formatting changes if needed. I too keep fighting with the IDE that wants to auto-format things. But it only creates unnecessary noise for reviewers :-(

thomaswoeckinger commented:
Pushed again, hope it is enough. Waiting for your suggestions...

janhoy commented:
There is stil a TON of unrelated whitespace changes in the patch. If you want to reformat code, best practice is to open another PR for that separately.

What I normally do to clean up is first disable the aggressive reformatting feature in the IDE, then open "Compare with branch ..." in my IDE (IntelliJ) and in the diff view, click restore on every single diff change that is messed up by the IDE, then commit and push those changes. You'll then see that your diff becomes MUCH smaller and easier to read!

martin-g commented:
One can easily hide the white space changes in Github UI by pressing the `No whitespace` button (or add manually `?w=1` to the url).

thomaswoeckinger commented:
Is it better for you if put the formatting in an seperate commit, i think
if i create a new pr this will never get merged, otherwise the format would
not be in the current state.

Jan Høydahl <notifications@github.com> schrieb am Do., 16. Mai 2019, 15:19:

> There is stil a TON of unrelated whitespace changes in the patch. If you
> want to reformat code, best practice is to open another PR for that
> separately.
>
> What I normally do to clean up is first disable the aggressive
> reformatting feature in the IDE, then open "Compare with branch ..." in my
> IDE (IntelliJ) and in the diff view, click restore on every single diff
> change that is messed up by the IDE, then commit and push those changes.
> You'll then see that your diff becomes MUCH smaller and easier to read!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/lucene-solr/pull/665?email_source=notifications&email_token=ACKCGQEYP7AEC5CH64ZIUB3PVVNPJA5CNFSM4HLKBL4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODVRY7YI#issuecomment-493064161>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACKCGQAVP6KGSZBUSHKBPRLPVVNPJANCNFSM4HLKBL4A>
> .
>

thomaswoeckinger commented:
I splitt up the changes into an additional commit, i think review is pretty easy now if you select next to last in the commit view.

Best,

Tom

janhoy commented:
It looks trivial to do a separate PR for SOLR-13347. Please do that if possible and let's avoid huge multi-issue patches.

I am still confused by unrelated formatting changes, consider opening a new JIRA and PR for those, I'm sure someone will take a look at it :) 

You might want to review the contributor guide: https://wiki.apache.org/solr/HowToContribute#Creating_the_patch_file

thomaswoeckinger commented:
Added PR for UUID only, will change this PR when it is merged, because unit test will fail otherwise

janhoy commented:
Thanks. I'm not saying I have the capacity or skills to commit these, but at least trying to help you shape your contributions into bite-sized parts that more likely attract committer attention. Once you remove all white-space edits and SOLR-1347 code from this PR it will be a lot easier to chew over.

noblepaul commented on solr/core/src/java/org/apache/solr/handler/loader/XMLLoader.java:
```diff
@@ -178,13 +183,12 @@ public void load(SolrQueryRequest req, SolrQueryResponse rsp, ContentStream stre
         }
-        parser = (charset == null) ?
-          inputFactory.createXMLStreamReader(is) : inputFactory.createXMLStreamReader(is, charset);
+        parser = (charset == null) ? inputFactory.createXMLStreamReader(is)
```
It is very hard to know if this is a real change or a formatting change , looking at the diff. Is it possible to get rid of the irrelevant changes? I'm all for formatting changes as a separate PR

noblepaul commented on solr/core/src/java/org/apache/solr/handler/loader/XMLLoader.java:
```diff
@@ -178,13 +183,12 @@ public void load(SolrQueryRequest req, SolrQueryResponse rsp, ContentStream stre
         }
-        parser = (charset == null) ?
-          inputFactory.createXMLStreamReader(is) : inputFactory.createXMLStreamReader(is, charset);
+        parser = (charset == null) ? inputFactory.createXMLStreamReader(is)
```
It is very hard to know if this is a real change or a formatting change , looking at the diff. Is it possible to get rid of the irrelevant changes? I'm all for formatting changes as a separate PR

thomaswoeckinger replied:
As i mentioned above if you select next to last in the commit view, it is easy to review, formatting is done only in the last commit

noblepaul commented on solr/core/src/java/org/apache/solr/schema/BoolField.java:
```diff
@@ -97,10 +98,9 @@ public boolean incrementToken() throws IOException {
-          if (ch==-1) return false;
+          if (ch == -1) return false;
           termAtt.copyBuffer(
-                  ((ch=='t' || ch=='T' || ch=='1') ? TRUE_TOKEN : FALSE_TOKEN)
```
is there anything changed here?

thomaswoeckinger replied:
As i mentioned above if you select next to last in the commit view, it is easy to review, formatting is done only in the last commit

thomaswoeckinger commented:
Will update this PR when #681 is merged, so @noblepaul and @janhoy plz have a look

thomaswoeckinger commented:
Push includes commit from #681, will rebase this PR after merge of #681

gerlowskija commented:
PR looks good to me.  Sorry for my tardiness in reviewing it.

I have some qualms about introducing a new test base class entirely as this PR does.  Our test-base situation is already quite jumbled and there's a lot of developer confusion around when each test-case should be used.  There's also been some recent movement towards randomizing the SolrJ RequestWriter in one of the existing test bases.  So it's hard to say whether adding a new test-base-class will hold up long term.  But I don't want to let the perfect get in the way of the good here - especially since this PR adds some awesome test coverage for atomic updates of a large variety of field types.

If Noble gets a chance to review #681 I'll circle back and merge this.  (Assuming he doesn't beat me to it).

gerlowskija commented:
I agree with Jan's complaints about IDE-triggered changes in this PR btw.  The `w=1` trick is a cool one to know, but reviews are still made tougher when there is noise from re-organized imports, `final` addition/removal, etc.

Look out for this in subsequent commits if you can.  If our ant targets produce IDE configs that trigger these reformattings as automatic save actions, we'll have to fix that.  Thanks for bringing it up.

ErickErickson commented:
Jason:

I don’t know enough about the “idea” target to know how to set this automagically, but that’s a great idea. Do you want to raise a JIRA?

I think I set this up globally somehow because no matter how often I execute the ant target it doesn’t reformat everything, just changed lines. But it’s trappy as it stands so +1 to changing  it.

Erick



> On Jun 12, 2019, at 8:38 AM, Jason Gerlowski <notifications@github.com> wrote:
> 
> PR looks good to me. Sorry for my tardiness in reviewing it.
> 
> I have some qualms about introducing a new test base class entirely as this PR does. Our test-base situation is already quite jumbled and there's a lot of developer confusion around when each test-case should be used. There's also been some recent movement towards randomizing the SolrJ RequestWriter in one of the existing test bases. So it's hard to say whether adding a new test-base-class will hold up long term. But I don't want to let the perfect get in the way of the good here - especially since this PR adds some awesome test coverage for atomic updates of a large variety of field types.
> 
> If Noble gets a chance to review #681 I'll circle back and merge this. (Assuming he doesn't beat me to it).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

thomaswoeckinger commented:
> PR looks good to me. Sorry for my tardiness in reviewing it.
> 
> I have some qualms about introducing a new test base class entirely as this PR does. Our test-base situation is already quite jumbled and there's a lot of developer confusion around when each test-case should be used. There's also been some recent movement towards randomizing the SolrJ RequestWriter in one of the existing test bases. So it's hard to say whether adding a new test-base-class will hold up long term. But I don't want to let the perfect get in the way of the good here - especially since this PR adds some awesome test coverage for atomic updates of a large variety of field types.
> 
> If Noble gets a chance to review #681 I'll circle back and merge this. (Assuming he doesn't beat me to it).

You are right about the number of existing test bases, but from my point of view SolrJettyTestBase is hiding the usage of EmbeddedSolrServer, which no one is expecting to be used there.

So i think it is clearer now.

To get rid of this situation the whole test base have to be re-written, which is to much for this PR ;-)

thomaswoeckinger commented:
Rebased and ready to review

janhoy commented:
Linking with SOLR-13539

thomaswoeckinger commented:
Any comments on this?

gerlowskija commented:
The code looks good to me.  I'm going to run tests and merge if all looks ok.

gerlowskija commented on solr/solrj/src/java/org/apache/solr/client/solrj/util/ClientUtils.java:
```diff
@@ -104,35 +103,51 @@ public static void writeXML( SolrInputDocument doc, Writer writer ) throws IOExc
       if (v != null) {
-        XML.writeXML(writer, "field", valWriter, "name", name);
+        if (binary) {
+          XML.writeXML(writer, "field", valWriter, "name", name, "dt", XMLLoader.BINARY_BASE64);
```
[-1] When I pull this code down locally, I get compilation errors.  XMLLoader lives in solr-core.  SolrJ doesn't depend on solr-core (it's the other way around), so when I try to compile this PR, I get:

```
    [javac] Compiling 716 source files to /Users/jasongerlowski/checkouts/lucene-solr/solr/build/solr-solrj/classes/java
    [javac] /Users/jasongerlowski/checkouts/lucene-solr/solr/solrj/src/java/org/apache/solr/client/solrj/util/ClientUtils.java:36: error: package org.apache.solr.handler.loader does not exist
    [javac] import org.apache.solr.handler.loader.XMLLoader;
```

We should move this constant to somewhere in SolrJ so that it can be seen in both solr-core and solr-solrj.

thomaswoeckinger replied:
This a result of the single project creation for eclipse, it is nearly impossible to see which one is depending on what ever, it is all in the same classpath, anyway i will take a look and push again

thomaswoeckinger replied:
In the same package there is a Constants class i can put it in there.

gerlowskija commented on solr/solrj/src/test/org/apache/solr/client/solrj/GetByIdTest.java:
```diff
@@ -27,13 +27,13 @@
 
-public class GetByIdTest extends SolrJettyTestBase {
-  
+public class GetByIdTest extends EmbeddedSolrServerTestBase {
```
[Q] Why are you changing the base class of this test here?  I'm fine with introducing a new base test-class in this PR, and maybe we want to eventually change other classes to use this new base.  But I'd rather that happen in a separate PR.

(That way, if tests start failing it's clear what's at fault.  With the changes bundled together like this it's harder to tell whether the bugfixes caused any hypothetical test failures, or the test-changes themselves are unsound.)

thomaswoeckinger replied:
Because the test uses the hidden EmbeddedSolrServer, so moved it to the new base

gerlowskija commented on solr/solrj/src/test/org/apache/solr/client/solrj/response/TestSpellCheckResponse.java:
```diff
@@ -27,21 +29,22 @@
  * @since solr 1.3
  */
-public class TestSpellCheckResponse extends SolrJettyTestBase {
+public class TestSpellCheckResponse extends EmbeddedSolrServerTestBase {
```
[Q] Why are you changing the base class of this test here?  I'm fine with introducing a new base test-class in this PR, and maybe we want to eventually change other classes to use this new base.  But I'd rather that happen in a separate PR.

(That way, if tests start failing it's clear what's at fault.  With the changes bundled together like this it's harder to tell whether the bugfixes caused any hypothetical test failures, or the test-changes themselves are unsound.)

thomaswoeckinger replied:
Because the test uses the hidden EmbeddedSolrServer, so moved it to the new base

gerlowskija commented on solr/solrj/src/test/org/apache/solr/client/solrj/response/TestSuggesterResponse.java:
```diff
@@ -32,14 +33,15 @@
  *
  */
-public class TestSuggesterResponse extends SolrJettyTestBase {
+public class TestSuggesterResponse extends EmbeddedSolrServerTestBase {
```
[Q] Ditto re: changing base class to EmbeddedSolrServerTestBase

thomaswoeckinger replied:
Because the test uses the hidden EmbeddedSolrServer, so moved it to the new base

gerlowskija commented on solr/test-framework/src/java/org/apache/solr/SolrJettyTestBase.java:
```diff
@@ -140,29 +139,17 @@ public synchronized SolrClient getSolrClient() {
+      final HttpSolrClient client = getHttpSolrClient(url, DEFAULT_CONNECTION_TIMEOUT);
+      return client;
+    } catch (final Exception ex) {
+      throw new RuntimeException(ex);
```
[-1] Either I missed this on an earlier review, or this change is new in the latest iteration of this PR.  (Publishing changes with git's "force-push" option as you've been doing makes it hard to see which changes are new and which have been around for a few iterations).

Anyway, I didn't notice this change previously, but some test failures drew my attention to it.

When I check out your PR and merge master into it locally, I get several test failures:

```
   [junit4] Tests with failures [seed: D067023E71C9E8C8]:
   [junit4]   - org.apache.solr.update.RootFieldTest.testUpdateWithChildDocs
   [junit4]   - org.apache.solr.update.RootFieldTest.testLegacyBlockProcessing
   [junit4]   - org.apache.solr.handler.tagger.EmbeddedSolrNoSerializeTest.testAssertTagStreamingWithSolrTaggerRequest
   [junit4]   - org.apache.solr.handler.tagger.EmbeddedSolrNoSerializeTest.testTag
   [junit4]   - org.apache.solr.update.processor.AbstractAtomicUpdatesMultivalueTest.initializationError
```

I looked into the RootFieldTest failure, and it is failing because of an NPE in this method:

```
   [junit4] ERROR   0.01s | RootFieldTest.testLegacyBlockProcessing <<<
   [junit4]    > Throwable #1: java.lang.RuntimeException: java.lang.NullPointerException
   [junit4]    > 	at __randomizedtesting.SeedInfo.seed([84EEE01E1A0D0DB8:E5D4F6A2DBEAF389]:0)
   [junit4]    > 	at org.apache.solr.SolrJettyTestBase.createNewSolrClient(SolrJettyTestBase.java:152)
   [junit4]    > 	at org.apache.solr.SolrJettyTestBase.getSolrClient(SolrJettyTestBase.java:136)
   [junit4]    > 	at org.apache.solr.update.RootFieldTest.testLegacyBlockProcessing(RootFieldTest.java:58)
   [junit4]    > 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
   [junit4]    > 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
   [junit4]    > 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
   [junit4]    > 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
   [junit4]    > 	at java.base/java.lang.Thread.run(Thread.java:834)
   [junit4]    > Caused by: java.lang.NullPointerException
   [junit4]    > 	at org.apache.solr.SolrJettyTestBase.createNewSolrClient(SolrJettyTestBase.java:148)
   [junit4]    > 	... 41 more
```

Was there a reason you removed the `jetty != null` check from this block, and the corresponding `else` block that created an EmbeddedSolrServer if jetty hadn't been called?  It seems like those are still needed here.

thomaswoeckinger replied:
See comments above

gerlowskija commented on solr/test-framework/src/java/org/apache/solr/SolrJettyTestBase.java:
```diff
@@ -178,13 +165,6 @@ public static void cleanUpJettyHome(File solrHome) throws Exception {
     }
   }
 
-  public static void initCore() throws Exception {
```
[Q] Why did you remove this?  Was it causing you an issue somewhere?

thomaswoeckinger commented:
> Hey Thomas, the code compiles and passes `precommit` now, which is a good step, but it looks like there are some test failures that this PR introduces. It looks like they're related to a change you made to SolrJettyTestBase. See one of my inline comments for more details. I also asked a few questions that I'd be interested in answers to before merging.
> 
No problem, code can only be getting better, and there are always reasons when questions are asked.

> Let me know if you're having any trouble reproducing those test failures locally.
> 
No troubles but as i mentioned before, working with the generated eclipse project is really a pain!

> Sorry that this has gone back and forth so many rounds. I imagine you're likely frustrated. But we're very close to getting this merged. Thanks for your patience and hard work on this.
You don't have to be sorry about that, on my side we are using solr server pretty much, so i am always interested in better test coverage an cleaner code base.

thomaswoeckinger commented:
Already had time to review?

thomaswoeckinger commented on solr/solrj/src/test/org/apache/solr/client/solrj/response/TermsResponseTest.java:
```diff
@@ -15,31 +15,33 @@
  */
-public class TermsResponseTest extends SolrJettyTestBase {
-
+public class TermsResponseTest extends EmbeddedSolrServerTestBase {
```
Because the test uses the hidden EmbeddedSolrServer, so moved it to the new base

gerlowskija commented:
Changes look good.  Running tests now and hoping to commit later today once those pass.

gerlowskija commented:
Hmm, I'm still seeing test-failures with this PR that look pretty related to the changes:

```
   [junit4] Tests with failures [seed: AD13CA5C9C515876]:
   [junit4]   - org.apache.solr.update.processor.AbstractAtomicUpdatesMultivalueTest.initializationError
   [junit4]   - org.apache.solr.cloud.TestCloudConsistency.testOutOfSyncReplicasCannotBecomeLeaderAfterRestart
   [junit4]   - org.apache.solr.handler.tagger.EmbeddedSolrNoSerializeTest.testTag
   [junit4]   - org.apache.solr.handler.tagger.EmbeddedSolrNoSerializeTest.testAssertTagStreamingWithSolrTaggerRequest
```

TestCloudConsistency might just be flaky, but the other ones seem relevant at least.  Are you able to see these locally, or is it just me?  To reproduce them, I'm checking out the latest `master`, and then cherry-picking your commits on top.  The commands look like:

```
git fetch -f origin pull/665/head:SOLR_13539_thomas # Checkout this PR branch locally
git checkout master
git pull # Make sure I've got latest commits on master
git checkout -b SOLR_13539
git cherry-pick fec20285eb02890043ececf9c71a1dbeaceaf051
git cherry-pick b378868143bf01cb3d2a53665ff10c564b593189
ant test
```

gerlowskija commented:
I was able to clean up some of the failures locally by removing the second clause in the if-statement below.  (This was added in this PR in EmbeddedSolrServer)

```
 private Set<ContentStream> getContentStreams(final SolrRequest<?> request) throws IOException {
    if (request.getMethod() == SolrRequest.METHOD.GET || request instanceof QueryRequest) return null;
```

This makes sense....there are some QueryRequest implementations that are supposed to be POST's.  (e.g. JSON-queries and JSON-faceting requests, Solr tagging requests, etc.).  But there are still other failing tests that rely on SolrJettyTestBase.  So I haven't figured it out yet, but something's still not quite right in that base-class.

With Cao Manh Dat working on the 8.1.2 release, and Ignacio Vera talking about an 8.2 shortly afterward, if we can't figure out the test issues soon, it might be best to split the functionality changes and test-base changes into two separate PR's if possible.  I'd hate for this fix to miss the next crop of releases because we couldn't cross the finish line on test-changes also bundled into this commit.

thomaswoeckinger commented:
> I was able to clean up some of the failures locally by removing the second clause in the if-statement below. (This was added in this PR in EmbeddedSolrServer)
> 
> ```
>  private Set<ContentStream> getContentStreams(final SolrRequest<?> request) throws IOException {
>     if (request.getMethod() == SolrRequest.METHOD.GET || request instanceof QueryRequest) return null;
> ```
> 
> This makes sense....there are some QueryRequest implementations that are supposed to be POST's. (e.g. JSON-queries and JSON-faceting requests, Solr tagging requests, etc.). But there are still other failing tests that rely on SolrJettyTestBase. So I haven't figured it out yet, but something's still not quite right in that base-class.
> 
> With Cao Manh Dat working on the 8.1.2 release, and Ignacio Vera talking about an 8.2 shortly afterward, if we can't figure out the test issues soon, it might be best to split the functionality changes and test-base changes into two separate PR's if possible. I'd hate for this fix to miss the next crop of releases because we couldn't cross the finish line on test-changes also bundled into this commit.

Had some problems with the generated ivy cache, but after all i managed to run all test successfully. The new PR is #720, so have a look, if merged i can update this one

gerlowskija commented:
Ok.  I was worried we'd miss 8.1.2, so I took your changes to AbstractEnumField, BinaryField, BoolField, and UUIDField and merged them in a much smaller commit of their own. (see here: https://github.com/apache/lucene-solr/commit/8242e6ce1da141e5585c0a8cd7ffc5953c5bf035)  This commit didn't include any of your test changes, unfortunately.

(I don't like to merge code without tests, but since I'd seen and run the tests you added in this PR specifically for those field types, and tested it manually, I felt pretty confident.)

Anyway, I'll take a look at 755 and run the tests later today.  Since your latest test changes are over in 755, and the atomic-update related field changes have already been merged, is there a reason to keep this PR around?  We could just work off of 755 and close this out, unless there's a reason you see to keep both 755 and 665 open?

thomaswoeckinger commented:
> Ok. I was worried we'd miss 8.1.2, so I took your changes to AbstractEnumField, BinaryField, BoolField, and UUIDField and merged them in a much smaller commit of their own. (see here: [8242e6c](https://github.com/apache/lucene-solr/commit/8242e6ce1da141e5585c0a8cd7ffc5953c5bf035)) This commit didn't include any of your test changes, unfortunately.
> 
> (I don't like to merge code without tests, but since I'd seen and run the tests you added in this PR specifically for those field types, and tested it manually, I felt pretty confident.)
> 
> Anyway, I'll take a look at 755 and run the tests later today. Since your latest test changes are over in 755, and the atomic-update related field changes have already been merged, is there a reason to keep this PR around? We could just work off of 755 and close this out, unless there's a reason you see to keep both 755 and 665 open?

The only reason is that the related issues are depending on each other, so i think it is still more clear to do this whole stuff with the PR's #755 and #665. Rebasing should be no problem anyway.

thomaswoeckinger commented:
Rebased against PR #755, resolved conflicts, tests are all running, ready to merge after #755 is merged

thomaswoeckinger commented:
Anything new?

thomaswoeckinger commented:
Still nothing?

thomaswoeckinger commented:
@gerlowskija Are you still working on this?

thomaswoeckinger commented:
Rebased again, @gerlowskija it is two weeks ago since last notice, maybe someone else could takeover the review if jason is to busy?

thomaswoeckinger commented:
> I was able to clean up some of the failures locally by removing the second clause in the if-statement below. (This was added in this PR in EmbeddedSolrServer)
> 
> ```
>  private Set<ContentStream> getContentStreams(final SolrRequest<?> request) throws IOException {
>     if (request.getMethod() == SolrRequest.METHOD.GET || request instanceof QueryRequest) return null;
> ```
> 
> This makes sense....there are some QueryRequest implementations that are supposed to be POST's. (e.g. JSON-queries and JSON-faceting requests, Solr tagging requests, etc.). But there are still other failing tests that rely on SolrJettyTestBase. So I haven't figured it out yet, but something's still not quite right in that base-class.
> 
> With Cao Manh Dat working on the 8.1.2 release, and Ignacio Vera talking about an 8.2 shortly afterward, if we can't figure out the test issues soon, it might be best to split the functionality changes and test-base changes into two separate PR's if possible. I'd hate for this fix to miss the next crop of releases because we couldn't cross the finish line on test-changes also bundled into this commit.

Reverted the if statement, reason was that internal code requires the whole content response to be null. Added DirectJsonQueryRequestFacetingEmbeddedTest to prove functionality.

A general comment, it would be great to switch to junit 5, test cases could be extracted to default interfaces and all test classes simply have to implement it and provide their solr client.

thomaswoeckinger commented:
Last interaction is now one month ago, so can anyone takeover the review please?

thomaswoeckinger commented:
So please, i put a lot of work into this, it would be really great if somebody could help to review and get this into the main line at least!

thomaswoeckinger commented on solr/core/src/java/org/apache/solr/handler/loader/XMLLoader.java:
```diff
@@ -429,7 +434,18 @@ public SolrInputDocument readDoc(XMLStreamReader parser) throws XMLStreamExcepti
           } else if ("field".equals(parser.getLocalName())) {
             // should I warn in some text has been found too
-            Object v = isNull ? null : text.toString();
+            Object v;
```
The interesting part is below, just using old style if then else for better readability, and yes it's required for binary support when using XML. The corresponding AtomicUpdateTest will fail otherwise

gerlowskija replied:
So, just to make sure I understand: what is "binary XML" here?  Is this XML where some node in the XML doc has binary content?  Is this normal XML where the whole XML doc/string has been encoded using some binary format for compression or quicker transmission?

> this was the reason binary XML support was not working at least since 6.6.2
Interesting.  Is this something that Solr claimed to support or had support for at some point?  Or this is a new ability that Solr has never had that you're adding here?

thomaswoeckinger replied:
> So, just to make sure I understand: what is "binary XML" here? Is this XML where some node in the XML doc has binary content? Is this normal XML where the whole XML doc/string has been encoded using some binary format for compression or quicker transmission?
> 
There are several ways to transport binary data over XML, i used the one which seems to be the clearest most readable one from my point of view, which is also the proposed way on xml.com: https://www.xml.com/pub/a/98/07/binary/binary.html

> > this was the reason binary XML support was not working at least since 6.6.2
> 
> Interesting. Is this something that Solr claimed to support or had support for at some point? Or this is a new ability that Solr has never had that you're adding here?

Some versions before 6.6.2 EmbeddedSolrServer used XMLCodec as default codec. When people start using it more in there test cases there where some issues regarding enums, etc.. so the codec was switched to JavaBinCodec which was the default when using SolrJ anyway. So this feature was not needed any more, anyway as far as i know all the other codecs are supporting transport of binary data. So if anyone want to use XML when talking to Solr he can send binaries now, it is also required for the new tests, the would fail otherwise.

gerlowskija replied:
Ah, I think I see.  Without this piece of the change, the new tests would fail because particular field types have binary data, and as-is we don't handle that correctly in SolrJ/EmbeddedSolrServer for wt=xml.

But figuring out how Solr should handle binary field data when using `wt=xml` is probably worth its own jira, for a few reasons:  (1) Other people in the community are likely to have opinions and want to chime in.  (2) It also, to be honest about my limitations, strays into areas of the code I don't know as well.  (3) It's fixing a different bug than these tests are meant for.

I'd like to see it get fixed, and I can work with you if you open a JIRA specifically for the binary-xml stuff.  I just don't think that fixing it here is the right approach.  Can you remove the binary-xml related changes, and either comment out the added tests that will fail, or add a TODO to add them once the underlying XML issue is fixed.  (If you file a separate JIRA and reference that in your TODO comment, others will have the context if they want it).

thomaswoeckinger replied:
> Ah, I think I see. Without this piece of the change, the new tests would fail because particular field types have binary data, and as-is we don't handle that correctly in SolrJ/EmbeddedSolrServer for wt=xml.
> 
Yes, of course!

> But figuring out how Solr should handle binary field data when using `wt=xml` is probably worth its own jira, for a few reasons: (1) Other people in the community are likely to have opinions and want to chime in. (2) It also, to be honest about my limitations, strays into areas of the code I don't know as well. (3) It's independent conceptually from the problem we started out trying to solve here (adding some atomic update tests to cover what we currently support).
> 
Added new issue SOLR-13762 and PR #883 

> I'd like to see it get fixed, and I can work with you if you open a JIRA specifically for the binary-xml stuff. I just don't think that fixing it here is the right approach. Can you remove the binary-xml related changes, and either comment out the added tests that will fail, or add a TODO to add them once the underlying XML issue is fixed. (If you file a separate JIRA and reference that in your TODO comment, others will have the context if they want it).

I was thinking about the same, so it was not much work anyway, and it is better separated now

thomaswoeckinger commented:
As i already mentioned on #755, any plans on merging this back to 8.x?

thomaswoeckinger commented:
One test is failing, testUserAndTestDefaultConfigsetsAreSame from TestConfigSetsAPI, have to investigate if it fails before or not

thomaswoeckinger commented:
Test is failing also before this commit, in both build environments ant, eclipse on my development host, so i ignored it locally. Does not seem to be related to this PR. All other tests are running successfully.

gerlowskija commented:
> As i already mentioned on #755, any plans on merging this back to 8.x?

Yep.  Generally committers are encouraged to let things "cook" on master awhile before merging things back to release branches like `branch_8x`.  This gives us more confidence that there's no weird flaky failures we've introduced.  It's also more convenient to wait to merge commits back to other branches once all related commits have been put on master.

I'm waiting mostly on the latter, since I've run the tests a good bit.  So once this gets merged (and I merge the patch that Tim Owen put on SOLR-13539), then I'll move everything back to `branch_8x` all at once.

thomaswoeckinger commented:
> > As i already mentioned on #755, any plans on merging this back to 8.x?
> 
> Yep. Generally committers are encouraged to let things "cook" on master awhile before merging things back to release branches like `branch_8x`. This gives us more confidence that there's no weird flaky failures we've introduced. It's also more convenient to wait to merge commits back to other branches once all related commits have been put on master.

Seems to be a good idea.
> 
> I'm waiting mostly on the latter, since I've run the tests a good bit. So once this gets merged (and I merge the patch that Tim Owen put on SOLR-13539), then I'll move everything back to `branch_8x` all at once.

Great



> > As i already mentioned on #755, any plans on merging this back to 8.x?
> 
> Yep. Generally committers are encouraged to let things "cook" on master awhile before merging things back to release branches like `branch_8x`. This gives us more confidence that there's no weird flaky failures we've introduced. It's also more convenient to wait to merge commits back to other branches once all related commits have been put on master.
> 
> I'm waiting mostly on the latter, since I've run the tests a good bit. So once this gets merged (and I merge the patch that Tim Owen put on SOLR-13539), then I'll move everything back to `branch_8x` all at once.

thomaswoeckinger commented:
Removed JavaDoc reference to EmbeddedSolrServer as suggested by @dsmiley

thomaswoeckinger commented:
@gerlowskija ready to merge?

thomaswoeckinger commented:
Rebased on the PR #883

thomaswoeckinger commented:
@gerlowskija: would be great if we can get this into 8.3, which will start in a about 2 weeks

gerlowskija commented:
Hi Thomas.  I'd really like to get this (#665) in for 8.3, but right now it's bundled to your change in #883.  I can't merge this without #883.

And I haven't had the time I need to understand how the different pieces of #883 (response formats, field types, binary content) fit together and what the best design is there.

So what I'm going to do is try to unbundle the two PRs, or at least flip the ordering.  I'll take what you've uploaded to #665 here and either comment out or remove entirely the tests that hit this binary-XML issue.  They can be added back in #883, once binary-xml works for these field types.
  
There might still be time to get #883 in, as I have a chunk of time now that I didn't before.  But even if we don't get to it for 8.3, it won't prevent #665 from getting in.

thomaswoeckinger commented:
> Hi Thomas. I'd really like to get this (#665) in for 8.3, but right now it's bundled to your change in #883. I can't merge this without #883.
> 
> And I haven't had the time I need to understand how the different pieces of #883 (response formats, field types, binary content) fit together and what the best design is there.
> 
> So what I'm going to do is try to unbundle the two PRs, or at least flip the ordering. I'll take what you've uploaded to #665 here and either comment out or remove entirely the tests that hit this binary-XML issue. They can be added back in #883, once binary-xml works for these field types.
> 
> There might still be time to get #883 in, as I have a chunk of time now that I didn't before. But even if we don't get to it for 8.3, it won't prevent #665 from getting in.
No problem i can comment out the binary test, rebase and push again.
Just a second!

thomaswoeckinger commented:
Pushed again without ignored binary test

thomaswoeckinger commented:
pre commit check is still toggling

gerlowskija commented:
> pre commit check is still toggling, it is working locally

Precommit has some known issues on master that make it a little flaky everywhere.  But it does seem like it has a higher rate of failure on Github (on all PR's...it's not specific to this one).  I wonder what that's about...

Running tests locally now.  Will merge later this morning assuming things check out.

thomaswoeckinger commented:
> > pre commit check is still toggling, it is working locally
> 
> Precommit has some known issues on master that make it a little flaky everywhere. But it does seem like it has a higher rate of failure on Github (on all PR's...it's not specific to this one). I wonder what that's about...
> 
> Running tests locally now. Will merge later this morning assuming things check out.

Great to hear.

thomaswoeckinger commented:
Any plans on getting this into 8.3?

thomaswoeckinger commented:
Thx Jason for moving it to 8.3
